File: frontend/lib/types/suggestions.ts
/**
 * Espejo manual de `backend/app/schemas/suggestions.py`.
 * Cambios en el schema Pydantic se replican aquí en el mismo commit
 * (skill `api-schema-sync`).
 */

import type { UUID } from "./auth";
import type { BankrollStrategy } from "./sessions";

/** Fuerza de la señal (§2.6). Describe evidencia pasada, nunca certeza futura. */
export type SignalStrength = "strong" | "medium" | "weak";

export interface StatisticalSuggestionItem {
  category: string;
  option_label: string;
  /** Estos dos se muestran SIEMPRE juntos (regla anti-falacia del jugador, §2). */
  theoretical_probability: number;
  observed_frequency_shrunk: number;
  /** Intervalo de Wilson al 95% sobre los conteos crudos (§2.2). */
  observed_ci_low: number;
  observed_ci_high: number;
  deviation: number;
  significance_score: number;
  strength: SignalStrength;
  ev: number;
  /** χ² ya corregido por comparaciones múltiples (§2.4). null si no está activo. */
  chi_square_pvalue_adjusted: number | null;
}

export interface StatisticalSuggestionsPanel {
  window_size_used: number;
  top: StatisticalSuggestionItem[];
  all_categories: Record<string, StatisticalSuggestionItem[]>;
}

export interface StreakAlert {
  category: string;
  option_label: string;
  consecutive_count: number;
  probability_of_streak: number;
}

export type BankrollAlertLevel = "info" | "caution" | "critical";

/** Alerta de gestión de banca: habla del dinero, nunca del próximo resultado. */
export interface BankrollAlertResponse {
  code: string;
  level: BankrollAlertLevel;
  message: string;
}

/** Dónde queda la progresión si el giro cierra en contra o a favor. Condicional. */
export interface NextStepResponse {
  stage: number;
  bet_per_sector: number;
  suggested_bet: number;
  /** Suponiendo que se apostó lo que pide la progresión. */
  bankroll_after: number;
  exceeds_table_limit: boolean;
  exceeds_bankroll: boolean;
  /** false si la sesión no tiene límite de pérdida. */
  reaches_loss_limit: boolean;
}

/**
 * Tamaño de apuesta que exige el escalón actual de la progresión.
 * Sugiere *cuánto* arriesgar, nunca a qué apostar ni qué resultado esperar.
 */
export interface BankrollSuggestionResponse {
  strategy: BankrollStrategy;
  /** Total del giro, sumando todos los sectores. */
  suggested_bet: number;
  /** En modo dos-sectores, la apuesta por docena/columna. */
  bet_per_sector: number;
  /** 1 en modo 1:1, 2 en modo dos-sectores. */
  sectors: number;
  stage: number;
  /** Perdido si fallaron todos los escalones hasta aquí. */
  cumulative_risked: number;
  /**
   * Con cuánto queda la serie completa si este giro se gana. En la progresión
   * de dos sectores es 0 del escalón 2 en adelante: recupera, no deja ganancia.
   */
  net_result_if_won: number;
  recovers_only_to_break_even: boolean;
  exceeds_table_limit: boolean;
  exceeds_bankroll: boolean;
  risk_warning: string | null;
  ruin_probability_estimate: number | null;
  /** Recordatorio fijo: ninguna progresión altera la ventaja de la casa (§2.8). */
  disclaimer: string;
  next_if_lost: NextStepResponse;
  next_if_won: NextStepResponse;
  /** Escalones seguidos, contando el actual, que la banca actual puede pagar. */
  stages_supported: number;
  /** Ordenadas de la más grave a la menos grave. */
  alerts: BankrollAlertResponse[];
}

/**
 * Una apuesta compatible con el modo de la estrategia de la sesión, con su
 * probabilidad teórica ya calculada por el motor: el frontend nunca la deriva.
 */
export interface EligibleBetResponse {
  id: string;
  label: string;
  category_id: string;
  group_ids: string[];
  theoretical_probability: number;
}

/** Una fila de la tabla de riesgo, con montos reales del usuario. */
export interface ProgressionRowResponse {
  stage: number;
  bet_per_sector: number;
  total_bet: number;
  cumulative_loss: number;
  exceeds_table_limit: boolean;
  exceeds_bankroll: boolean;
}

/**
 * Tabla que se muestra ANTES de activar una estrategia (§2.8): el riesgo debe
 * ser visible en pesos, no abstracto.
 */
export interface ProgressionTableResponse {
  strategy: BankrollStrategy;
  base_bet: number;
  sectors: number;
  rows: ProgressionRowResponse[];
  max_affordable_stages: number;
  ruin_probability_estimate: number | null;
  disclaimer: string;
}

export interface SuggestionSnapshot {
  spin_id: UUID;
  spin_index: number;
  created_at: string;
  panel: StatisticalSuggestionsPanel;
  bankroll: BankrollSuggestionResponse | null;
}

// ============================================================================
// Motor de recomendación (Fase 3, §2.10)
// ============================================================================
//
// La pantalla principal muestra esto: un mercado, un `signal_score` y su banda.
// Las frecuencias y probabilidades siguen viajando en la respuesta —regla
// anti-falacia del jugador, §2— pero se pintan en "¿Por qué recomienda esto?".

/** Qué hacer en el giro siguiente. */
export type RecommendationDecision = "RECOMMEND" | "NO_BET";

/**
 * Por qué el motor no recomienda apostar: todavía no hay giros suficientes para
 * medir, o los hay y ningún mercado llega al umbral.
 */
export type NoBetReason = "insufficient_data" | "below_threshold";

/**
 * Estado de salida del motor según los tres umbrales (§2.10):
 * - `none`: SIN SEÑAL — por debajo de `weak_threshold`. No apostar.
 * - `weak`: SEÑAL DÉBIL — de `weak_threshold` a `threshold`. Solo apuesta base.
 * - `medium`: SEÑAL MEDIA — de `threshold` a `strong_threshold`.
 * - `strong`: SEÑAL FUERTE — desde `strong_threshold`.
 * Describe la fuerza del criterio interno sobre la muestra ya ocurrida, no la
 * probabilidad de acertar el próximo giro.
 */
export type SignalBand = "none" | "weak" | "medium" | "strong";

/** Un NO APOSTAR se queda en `PENDING`: no hubo nada que acertar ni que fallar. */
export type RecommendationOutcome = "PENDING" | "HIT" | "MISS";

/** Lo que el mercado hizo dentro de una ventana (10, 20, 50, 100 giros). */
export interface WindowStatResponse {
  window: number;
  spins_used: number;
  /** Regla anti-falacia del jugador (§2): estos dos nunca viajan por separado. */
  theoretical_probability: number;
  observed_frequency_shrunk: number;
  raw_count: number;
  /** Intervalo de Wilson al 95% sobre los conteos crudos (§2.2). */
  observed_ci_low: number;
  observed_ci_high: number;
  deviation: number;
  /** Desviación estandarizada con signo. Positiva = salió más de lo esperado. */
  z: number;
}

/** De dónde sale cada punto del `signal_score`. */
export interface ScoreComponentsResponse {
  /** D: z de la ventana más larga, normalizado a [0,1]. */
  deviation: number;
  /** R: z de la ventana más corta. */
  recency: number;
  /**
   * C: coherencia entre tramos disjuntos del historial. null con un solo tramo
   * — un tramo no tiene con qué ser consistente, y su peso pasa a D y R.
   */
  consistency: number | null;
  weight_deviation: number;
  weight_recency: number;
  weight_consistency: number;
  /** Puntos del respaldo del χ². Solo con ≥200 giros y p corregido < 0.05. */
  chi_square_bonus: number;
}

/** Una alternativa del catálogo de mercados de la variante. */
export interface MarketResponse {
  key: string;
  label: string;
  category_id: string;
  group_ids: string[];
  /** Cuántas zonas cubre a la vez: 1 (negro) o 2 (dos docenas). */
  sectors: number;
  /** Resultados cubiertos. Cambia con la variante: 24/37 vs. 24/38. */
  coverage: number;
  payout: number;
}

export interface ScoredMarketResponse {
  market: MarketResponse;
  signal_score: number;
  signal_band: SignalBand;
  components: ScoreComponentsResponse;
  windows: WindowStatResponse[];
  chi_square_pvalue_adjusted: number | null;
}

/**
 * Lo que una progresión pide para el mercado recomendado. `applicable: false`
 * no es un error: es la respuesta honesta cuando la progresión no encaja con la
 * forma del mercado (recuperación de dos sectores sobre un mercado de uno).
 */
export interface MarketStakeResponse {
  strategy: BankrollStrategy;
  applicable: boolean;
  reason: string | null;
  stage: number;
  bet_per_sector: number;
  total_bet: number;
  sectors: number;
  cumulative_risked: number;
  net_result_if_won: number;
  recovers_only_to_break_even: boolean;
  exceeds_bankroll: boolean;
  exceeds_table_limit: boolean;
}

/**
 * La recomendación para el giro siguiente. `best` viaja también con `NO_BET`:
 * siempre hay un mejor candidato, sólo que por debajo del umbral.
 */
export interface RecommendationResponse {
  session_id: UUID;
  decision: RecommendationDecision;
  /** null con RECOMMEND. */
  no_bet_reason: NoBetReason | null;
  /** Giros por debajo de los cuales un NO_BET es por falta de información. */
  min_spins_for_signal: number;
  /** Umbral medio: desde aquí la señal es MEDIA. */
  threshold: number;
  /** Umbral débil: desde aquí hay recomendación (SEÑAL DÉBIL, solo apuesta base). */
  weak_threshold: number;
  /** Umbral alto: desde aquí la señal es FUERTE. */
  strong_threshold: number;
  signal_score: number;
  signal_band: SignalBand;
  /** El mercado a apostar. null con NO_BET. */
  market: MarketResponse | null;
  best: ScoredMarketResponse | null;
  /** Las tres progresiones sobre el mercado. Vacío con NO_BET. */
  stakes: MarketStakeResponse[];
  /** Todos los mercados evaluados, ya ordenados por el desempate. */
  candidates: ScoredMarketResponse[];
  window_size_used: number;
  total_spins: number;
  /** Recordatorio fijo al pie de la tarjeta (§2.10). */
  disclaimer: string;
}

/** Una recomendación ya emitida y su cierre — historial y auto-evaluación. */
export interface RecommendationRecord {
  id: UUID;
  session_id: UUID;
  spin_id: UUID | null;
  decision: RecommendationDecision;
  market_key: string;
  category: string;
  option_label: string;
  signal_score: number;
  signal_band: SignalBand;
  theoretical_probability: number;
  observed_frequency_shrunk: number;
  deviation: number;
  observed_ci_low: number;
  observed_ci_high: number;
  ev: number;
  chi_square_pvalue_adjusted: number | null;
  /** Monto en centavos, nunca float. null con NO_BET. */
  stake_cents: number | null;
  currency: string;
  outcome: RecommendationOutcome;
  resolved_spin_id: UUID | null;
  window_size_used: number;
  created_at: string; // ISO 8601
}

// ---------- Métricas internas del motor (solo admin, §2.10) ----------
//
// No se muestran al cliente: son el control de honestidad de §9 del comparativo,
// el que dice si el motor aporta información útil o sólo describe el pasado.

/** Sobre qué historiales corre el backtest. */
export type BacktestSource = "sessions" | "simulated";

export interface BacktestTally {
  recommendations: number;
  hits: number;
  misses: number;
  hit_rate: number;
  /** Resultado neto en unidades de apuesta base, con los pagos reales. */
  units: number;
  /** Resultado por unidad arriesgada. En una mesa sin sesgo tiende a −house_edge. */
  roi: number;
  /** Peor caída desde un pico previo, en unidades. */
  max_drawdown: number;
}

export interface BacktestBandRow extends BacktestTally {
  band: SignalBand;
  /** Cuántos NO APOSTAR salieron con esta banda en el mejor candidato. */
  no_bets: number;
}

export interface BacktestReport {
  source: string;
  sessions: number;
  spins_evaluated: number;
  decisions: number;
  recommendations: number;
  no_bets: number;
  no_bet_rate: number;
  /** Umbral medio (SEÑAL MEDIA) aplicado. */
  threshold: number;
  /** Umbral débil (mínimo para recomendar) aplicado. */
  weak_threshold: number;
  overall: BacktestTally;
  by_band: BacktestBandRow[];
  /** EV de cualquier apuesta en esta variante: la referencia del ROI. */
  house_edge_reference: number;
}

File: backend/app/schemas/suggestions.py
"""
Schemas: Statistical suggestions & Bankroll suggestions
"""
from datetime import datetime
from enum import Enum
from uuid import UUID
from typing import Optional
from pydantic import BaseModel, ConfigDict

from app.schemas.sessions import BankrollStrategy


class SignalStrength(str, Enum):
    """Fuerza de la señal según §2.6 — nunca "confianza": describe la evidencia
    pasada de la muestra, no una certeza sobre el próximo giro."""
    strong = "strong"
    medium = "medium"
    weak = "weak"


class StatisticalSuggestionItem(BaseModel):
    category: str
    option_label: str
    # Regla anti-falacia del jugador (§2): estos dos viajan y se muestran
    # SIEMPRE juntos, nunca uno solo.
    theoretical_probability: float
    # El sufijo `_shrunk` es deliberado: aquí va la frecuencia con shrinkage
    # bayesiano (§2.2), nunca la frecuencia cruda.
    observed_frequency_shrunk: float
    # Intervalo de Wilson al 95% sobre los conteos crudos (§2.2). Es la escala
    # del ruido: 6 de 10 y 600 de 1000 son ambos 60% y no dicen lo mismo.
    # Se muestra, no se interpreta: no hay un campo "esto se distingue del azar"
    # a propósito (serían 13 pruebas simultáneas — ver §2.2 del doc).
    observed_ci_low: float
    observed_ci_high: float
    deviation: float
    significance_score: float
    strength: SignalStrength
    ev: float  # §2.1: visible siempre junto a la señal
    # p-valor del χ² YA corregido por comparaciones múltiples (Benjamini-Hochberg,
    # §2.4). Es el que sostiene la decisión de activar la señal; el crudo no viaja
    # a la UI porque leerlo como significancia sobreestima lo que la familia de
    # pruebas respalda. null si el χ² no está activo para esa categoría.
    chi_square_pvalue_adjusted: Optional[float] = None


class StatisticalSuggestionsPanel(BaseModel):
    window_size_used: int
    top: list[StatisticalSuggestionItem] = []          # top 3 por significance_score
    all_categories: dict[str, list[StatisticalSuggestionItem]] = {}  # vista completa


class StreakAlert(BaseModel):
    category: str
    option_label: str
    consecutive_count: int
    probability_of_streak: float


class BankrollAlertLevel(str, Enum):
    info = "info"
    caution = "caution"
    critical = "critical"


class BankrollAlertResponse(BaseModel):
    """Alerta de gestión de banca: habla del dinero, nunca del próximo resultado."""
    code: str
    level: BankrollAlertLevel
    message: str


class NextStepResponse(BaseModel):
    """Dónde queda la progresión si el giro cierra en contra o a favor.

    Es condicional: describe los dos casos, nunca cuál va a ocurrir.
    """
    stage: int
    bet_per_sector: float
    suggested_bet: float
    bankroll_after: float         # suponiendo que se apostó lo que pide la progresión
    exceeds_table_limit: bool
    exceeds_bankroll: bool
    reaches_loss_limit: bool      # False si la sesión no tiene límite de pérdida


class BankrollSuggestionResponse(BaseModel):
    """Tamaño de apuesta que exige el escalón actual de la progresión.

    Sugiere *cuánto* arriesgar según la progresión que el usuario eligió; nunca
    a qué apostar ni qué resultado esperar.
    """
    strategy: BankrollStrategy
    suggested_bet: float          # total del giro, sumando todos los sectores
    bet_per_sector: float         # en modo dos-sectores, la apuesta por docena/columna
    sectors: int                  # 1 en modo 1:1, 2 en modo dos-sectores
    stage: int
    cumulative_risked: float      # perdido si fallaron todos los escalones hasta aquí
    # Con cuánto queda la serie completa si este giro se gana. En la progresión
    # de dos sectores es 0 del escalón 2 en adelante: recupera, no deja ganancia.
    net_result_if_won: float
    recovers_only_to_break_even: bool
    exceeds_table_limit: bool
    exceeds_bankroll: bool
    risk_warning: Optional[str] = None
    ruin_probability_estimate: Optional[float] = None
    # Recordatorio fijo: ninguna progresión altera la ventaja de la casa (§2.8).
    disclaimer: str
    next_if_lost: NextStepResponse
    next_if_won: NextStepResponse
    # Escalones seguidos, contando el actual, que la banca actual puede pagar.
    stages_supported: int
    # Ordenadas de la más grave a la menos grave.
    alerts: list[BankrollAlertResponse] = []


class EligibleBetResponse(BaseModel):
    """Una apuesta compatible con el modo de la estrategia de la sesión.

    Existe para que la interfaz pueda preguntar "¿sobre qué apuesta calculo el
    riesgo?" sin recalcular ninguna probabilidad por su cuenta: la teórica la
    entrega el motor.
    """
    id: str
    label: str
    category_id: str
    group_ids: list[str]
    theoretical_probability: float


class ProgressionRowResponse(BaseModel):
    """Una fila de la tabla de riesgo, con montos reales del usuario."""
    stage: int
    bet_per_sector: float
    total_bet: float
    cumulative_loss: float
    exceeds_table_limit: bool
    exceeds_bankroll: bool


class ProgressionTableResponse(BaseModel):
    """Tabla que se muestra ANTES de activar una estrategia (§2.8).

    El riesgo debe ser visible y no abstracto: el usuario ve en pesos lo que
    cuesta cada escalón antes de comprometerse con la progresión.
    """
    strategy: BankrollStrategy
    base_bet: float
    sectors: int
    rows: list[ProgressionRowResponse]
    max_affordable_stages: int
    ruin_probability_estimate: Optional[float] = None
    disclaimer: str


class SuggestionSnapshot(BaseModel):
    """Para GET /sessions/:id/suggestions/history"""
    spin_id: UUID
    spin_index: int
    created_at: datetime
    panel: StatisticalSuggestionsPanel
    bankroll: Optional[BankrollSuggestionResponse] = None


# ============================================================================
# Motor de recomendación (Fase 3, §2.10)
# ============================================================================
#
# Lo que la pantalla principal muestra es esto: un mercado, un `signal_score` y
# su banda. Las frecuencias y probabilidades siguen viajando en la respuesta
# —regla anti-falacia del jugador, §2— pero se pintan en la sección desplegable
# "¿Por qué recomienda esto?", no en la tarjeta principal.


class RecommendationDecision(str, Enum):
    """Qué hacer en el giro siguiente."""
    recommend = "RECOMMEND"
    no_bet = "NO_BET"


class NoBetReason(str, Enum):
    """Por qué el motor no recomienda apostar. null cuando sí recomienda."""
    # Menos giros que la ventana más corta: todavía no hay con qué medir.
    insufficient_data = "insufficient_data"
    # Hay datos, pero ningún mercado llega al umbral.
    below_threshold = "below_threshold"


class SignalBand(str, Enum):
    """Estado de salida del motor según los dos umbrales de §2.10:

    - `weak`: SIN SEÑAL — por debajo del umbral mínimo (`threshold`). No apostar.
    - `medium`: SEÑAL MEDIA — del umbral mínimo al alto (`strong_threshold`).
    - `strong`: SEÑAL FUERTE — desde el umbral alto.

    Describe la fuerza del criterio interno sobre la muestra ya ocurrida. No es
    la probabilidad de acertar el próximo giro — esa sigue siendo la teórica.
    """
    weak = "weak"
    medium = "medium"
    strong = "strong"


class RecommendationOutcome(str, Enum):
    """Cómo cerró una recomendación contra el giro siguiente.

    Un NO APOSTAR se queda en `PENDING`: no hubo nada que acertar ni que fallar.
    """
    pending = "PENDING"
    hit = "HIT"
    miss = "MISS"


class WindowStatResponse(BaseModel):
    """Lo que el mercado hizo dentro de una ventana (10, 20, 50, 100 giros)."""
    window: int
    spins_used: int
    # Regla anti-falacia del jugador (§2): estos dos nunca viajan por separado.
    theoretical_probability: float
    observed_frequency_shrunk: float
    raw_count: int
    # Intervalo de Wilson al 95% sobre los conteos crudos (§2.2): la escala del
    # ruido. Se muestra, no se interpreta.
    observed_ci_low: float
    observed_ci_high: float
    deviation: float
    # Desviación estandarizada con signo. Positiva = salió más de lo esperado.
    z: float


class ScoreComponentsResponse(BaseModel):
    """De dónde sale cada punto del `signal_score`.

    Se expone entero para que "¿Por qué recomienda esto?" pueda mostrar el
    desglose en vez de un número sin origen.
    """
    deviation: float            # D: z de la ventana más larga, normalizado a [0,1]
    recency: float              # R: z de la ventana más corta
    # C: coherencia entre tramos disjuntos del historial. null cuando hay un solo
    # tramo — un tramo no tiene con qué ser consistente, y entonces su peso se
    # reparte entre D y R.
    consistency: Optional[float] = None
    weight_deviation: float
    weight_recency: float
    weight_consistency: float
    # Puntos que suma el respaldo del χ². Solo con ≥200 giros y p corregido por
    # Benjamini-Hochberg < 0.05 (§2.4).
    chi_square_bonus: float


class MarketResponse(BaseModel):
    """Una alternativa del catálogo de mercados de la variante."""
    key: str
    label: str
    category_id: str
    group_ids: list[str]
    # Cuántas zonas cubre a la vez: 1 (negro) o 2 (dos docenas).
    sectors: int
    # Cuántos resultados cubre. Cambia con la variante: 24/37 europea, 24/38
    # americana para dos docenas.
    coverage: int
    payout: float


class ScoredMarketResponse(BaseModel):
    market: MarketResponse
    signal_score: float
    signal_band: SignalBand
    components: ScoreComponentsResponse
    windows: list[WindowStatResponse]
    chi_square_pvalue_adjusted: Optional[float] = None


class MarketStakeResponse(BaseModel):
    """Lo que una progresión pide para el mercado recomendado.

    `applicable=false` no es un error: es la respuesta honesta cuando la
    progresión no encaja con la forma del mercado (la recuperación de dos
    sectores sobre un mercado de un solo sector).
    """
    strategy: BankrollStrategy
    applicable: bool
    reason: Optional[str] = None
    stage: int
    bet_per_sector: float
    total_bet: float
    sectors: int
    cumulative_risked: float
    net_result_if_won: float
    recovers_only_to_break_even: bool
    exceeds_bankroll: bool
    exceeds_table_limit: bool


class RecommendationResponse(BaseModel):
    """La recomendación para el giro siguiente.

    `best` viaja también con `NO_BET`: siempre hay un mejor candidato, sólo que
    por debajo del umbral. Es lo que permite explicar por qué el motor calló.
    """
    session_id: UUID
    decision: RecommendationDecision
    # null con RECOMMEND.
    no_bet_reason: Optional[NoBetReason] = None
    # Giros por debajo de los cuales un NO_BET es por falta de información.
    min_spins_for_signal: int
    # Umbral mínimo: desde aquí hay recomendación (SEÑAL MEDIA).
    threshold: float
    # Umbral alto: desde aquí la señal es FUERTE.
    strong_threshold: float
    signal_score: float
    signal_band: SignalBand
    # El mercado a apostar. null con NO_BET.
    market: Optional[MarketResponse] = None
    # El mejor candidato, haya o no recomendación.
    best: Optional[ScoredMarketResponse] = None
    # Las tres progresiones sobre el mercado recomendado, en orden de menú.
    # Vacío con NO_BET: no hay monto que calcular.
    stakes: list[MarketStakeResponse] = []
    # Todos los mercados evaluados, ya ordenados por el desempate. Alimenta la
    # sección desplegable.
    candidates: list[ScoredMarketResponse] = []
    window_size_used: int
    total_spins: int
    # Recordatorio fijo al pie de la tarjeta (§2.10).
    disclaimer: str


class RecommendationRecord(BaseModel):
    """Una recomendación ya emitida y su cierre — historial y auto-evaluación."""
    id: UUID
    session_id: UUID
    spin_id: Optional[UUID] = None
    decision: RecommendationDecision
    market_key: str
    category: str
    option_label: str
    signal_score: float
    signal_band: SignalBand
    theoretical_probability: float
    observed_frequency_shrunk: float
    deviation: float
    observed_ci_low: float
    observed_ci_high: float
    ev: float
    chi_square_pvalue_adjusted: Optional[float] = None
    # Monto en centavos, nunca float. null con NO_BET.
    stake_cents: Optional[int] = None
    currency: str
    outcome: RecommendationOutcome
    resolved_spin_id: Optional[UUID] = None
    window_size_used: int
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


# ---------- Métricas internas del motor (solo admin, §2.10) ----------
#
# No se muestran al cliente: son el control de honestidad de §9 del comparativo,
# el que dice si el motor aporta información útil o sólo describe el pasado.


class BacktestTally(BaseModel):
    """Resultado acumulado de un grupo de recomendaciones."""
    recommendations: int
    hits: int
    misses: int
    hit_rate: float
    # Resultado neto en unidades de apuesta base, con los pagos reales
    # (1:1, 2:1, y 1:2 para dos docenas).
    units: float
    # Resultado por unidad arriesgada. En una mesa sin sesgo tiende a −house_edge.
    roi: float
    # Peor caída desde un pico previo, en unidades.
    max_drawdown: float


class BacktestBandRow(BacktestTally):
    band: SignalBand
    # Cuántos NO APOSTAR salieron con esta banda en el mejor candidato.
    no_bets: int


class BacktestReport(BaseModel):
    """Informe del backtest sobre historiales fuera de calibración."""
    source: str
    sessions: int
    spins_evaluated: int
    decisions: int
    recommendations: int
    no_bets: int
    no_bet_rate: float
    threshold: float
    overall: BacktestTally
    by_band: list[BacktestBandRow]
    # EV de cualquier apuesta en esta variante: la referencia contra la que se
    # lee el ROI. Un ROI cercano a esto es el resultado esperado, no un fallo.
    house_edge_reference: float
